src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1095,6 +1104,31 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
 
                     // Eat exception and try again.
                 }
+                finally
+                {
+                    Debug.Assert(http11ConnectionWaiter == null || http2ConnectionWaiter == null);
+                    CancelWithGlobalTimeout(this, http11ConnectionWaiter?.ConnectionCancellationTokenSource);
+                    CancelWithGlobalTimeout(this, http2ConnectionWaiter?.ConnectionCancellationTokenSource);
+                }
+            }
+
+            static void CancelWithGlobalTimeout(HttpConnectionPool pool, CancellationTokenSource? cts)
+            {
+                if (cts == null) return;
+                if (NetEventSource.Log.IsEnabled())
+                    pool.Trace($"Cancelling a pending connection attempt with timeout of {GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion} ms");
+
+                try
+                {
+                    if (GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion > 0)
+                        cts.CancelAfter(GlobalHttpSettings.SocketsHttpHandler.PendingConnectionTimeoutOnRequestCompletion);
+                    else
+                        cts.Cancel(); // Cancel immediately by default
+                }
+                catch (ObjectDisposedException)
+                {
+                    // There is a low-probability race for the CTS being disposed by the time of the cancellation call, ignore this ODE.
```

Comment:
It's not thread-safe to CTS.Cancel{After} and CTS.Dispose concurrently.  Further, if you Dispose it after having called CancelAfter, you won't end up canceling it as you'll be disposing of the timer that would have eventually done so.